### PR TITLE
Remove jQuery from Engelsystem scripts

### PR DESCRIPTION
**Pull request by weeman1337**
Replacement of select2 follows in a separate PR.

### CODE REVIEW

File: resources/assets/js/ready.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export const ready = (fn) => {
+  if (document.readyState !== 'loading') {
+    fn();
+  } else {
+    document.addEventListener('DOMContentLoaded', fn);
+  }
+}
```
Comment (ximex) [inline resources/assets/js/ready.js:7]: i would use the code from jquery
https://github.com/jquery/jquery/blob/main/src/core/ready.js#L57-L78
```js
function ready(callback) {
  // The ready event handler and self cleanup method
  function completed() {
    document.removeEventListener('DOMContentLoaded', completed);
    window.removeEventListener('load', completed);
    callback();
  }

  // Catch cases where onDocumentLoaded() is called after the browser event has already occurred.
  if (document.readyState !== 'loading') {
    // Handle it asynchronously to allow scripts the opportunity to delay ready
    window.setTimeout(callback);
  } else {
    // Use the handy event callback
    document.addEventListener('DOMContentLoaded', completed);

    // A fallback to window.onload, that will always work
    window.addEventListener('load', completed);
  }
}
```

**Comment by ximex** [comment]
maybe also integrate this changes in your PR: https://github.com/engelsystem/engelsystem/pull/1003

File: resources/assets/js/countdown.js [hunk 2]
```diff
@@ -67,14 +69,13 @@ function formatFromNow(timestamp) {
 /**
  * Initialises all countdown fields on the page.
  */
-$(function () {
-  $.each($('[data-countdown-ts]'), function (i, e) {
-    const span = $(e);
-    const timestamp = span.data('countdown-ts');
-    const text = span.html();
-    span.html(text.replace('%c', formatFromNow(timestamp)));
+ready(function () {
```
Comment (ximex) [inline resources/assets/js/countdown.js:72]: arrow-function?

File: resources/assets/js/countdown.js [hunk 3]
```diff
@@ -67,14 +69,13 @@ function formatFromNow(timestamp) {
...
+  document.querySelectorAll('[data-countdown-ts]').forEach((element) => {
+    const timestamp = element.dataset.countdownTs;
+    const template = element.innerHTML;
+    element.innerHTML = template.replace('%c', formatFromNow(timestamp));
     setInterval(function () {
```
Comment (ximex) [inline resources/assets/js/countdown.js:77]: arrow-function?

File: resources/assets/js/forms.js [hunk 4]
```diff
@@ -85,41 +97,49 @@ global.setHours = (hours) => {
   setInput(from, to);
 };
 
-$(function () {
+ready(function () {
```
Comment (ximex) [inline resources/assets/js/forms.js:100]: arrow-function?

File: resources/assets/js/forms.js [hunk 5]
```diff
@@ -85,41 +97,49 @@ global.setHours = (hours) => {
...
 });
 
 /*
  * Button to set current time in time input fields.
  */
-$(function () {
-  $('.input-group.time').each(function () {
-    const elem = $(this);
-    elem.find('button').on('click', function () {
+ready(function () {
```
Comment (ximex) [inline resources/assets/js/forms.js:118]: arrow-function?

File: resources/assets/js/forms.js [hunk 6]
```diff
@@ -85,41 +97,49 @@ global.setHours = (hours) => {
...
+          daySelector.value = dayElement.value;
           return false;
         }
       });
     });
   });
 });
 
-$(function () {
+ready(function () {
```
Comment (ximex) [inline resources/assets/js/forms.js:142]: arrow-function?

File: resources/assets/js/forms.js [hunk 7]
```diff
@@ -128,15 +148,23 @@ $(function () {
 /**
  * Show oauth buttons on welcome title click
  */
-$(function () {
-  $('#welcome-title').on('click', function () {
-    $('.btn-group.btn-group .btn.d-none').removeClass('d-none');
+ready(function () {
```
Comment (ximex) [inline resources/assets/js/forms.js:151]: arrow-function?

File: resources/assets/js/sticky-headers.js [hunk 8]
```diff
@@ -1,32 +1,44 @@
...
-        });
-        headers.css({
-          'top': Math.max(0, $(window).scrollTop() - top
-                        - 13
-                        + topReference.offset().top)
-                        + 'px'
-        });
-      });
-  }
+ready(function () {
```
Comment (ximex) [inline resources/assets/js/sticky-headers.js:17]: arrow-function?

File: resources/assets/js/vendor.js [hunk 9]
```diff
@@ -4,7 +4,3 @@ window.bootstrap = require('bootstrap');
 require('./forms');
 require('./sticky-headers');
 require('./countdown');
```
Comment (ximex) [inline resources/assets/js/vendor.js:6]: `import` instead of `require`?

**Comment by weeman1337** [comment]
> maybe also integrate this changes in your PR: #1003

A PR should target one thing. So this should keep being a separate PR.

File: resources/assets/js/countdown.js (hunk 2 above)
Comment (weeman1337) [reply resources/assets/js/countdown.js:72]: Mention this one time per PR is sufficient :wink:

File: resources/assets/js/forms.js (hunk 6 above)
Comment (weeman1337) [reply resources/assets/js/forms.js:142]: Mention this one time per PR is sufficient :wink:

File: resources/assets/js/forms.js (hunk 7 above)
Comment (weeman1337) [reply resources/assets/js/forms.js:151]: Mention this one time per PR is sufficient :wink:

File: resources/assets/js/sticky-headers.js (hunk 8 above)
Comment (weeman1337) [reply resources/assets/js/sticky-headers.js:17]: Mention this one time per PR is sufficient :wink:

File: resources/assets/js/ready.js (hunk 1 above)
Comment (weeman1337) [reply resources/assets/js/ready.js:7]: arrow-function? :trollface:

Comment (weeman1337) [reply resources/assets/js/ready.js:7]: `DOMContentLoaded` is supported by all current browsers. Fallback options are included in jQuery to support IE <= 8. I would mandate against copying over jQuery legacy stuff.

**Comment by weeman1337** [comment]
> ~fixes:~ relates to #1001

select2 migration still needs to be done. So it doesn't fix the issue, but it is a step in this direction.

File: resources/assets/js/ready.js (hunk 1 above)
Comment (ximex) [reply resources/assets/js/ready.js:7]: ok but i would rename `fn` to `callback`.
And should we do this with the `window.setTimeout(callback);`?

File: resources/assets/js/vendor.js (hunk 9 above)
Comment (weeman1337) [reply resources/assets/js/vendor.js:6]: Out of scope for this PR. https://github.com/engelsystem/engelsystem/issues/1006

File: resources/assets/js/ready.js (hunk 1 above)
Comment (weeman1337) [reply resources/assets/js/ready.js:7]: > ok but i would rename fn to callback.

:man_shrugging: can do

> And should we do this with the window.setTimeout(callback);?

Any specific reason for this?

Comment (ximex) [reply resources/assets/js/ready.js]: i only thought if jquery did it they would have a reason for it.
i think this is that in both cases the callback gets pushed at the end of the event loop (and not get executed directly) and behaves the same way

Comment (weeman1337) [reply resources/assets/js/ready.js]: jQuery is from the very past :slightly_smiling_face: The snippet above is already battle proven.

File: resources/assets/js/forms.js [hunk 10]
```diff
@@ -85,41 +97,49 @@ global.setHours = (hours) => {
...
-      const input = elem.children('input').first();
-      input.val(formatTime(now));
-      const daySelector = $('#' + input.attr('id').replace('time', 'day'));
-      const days = daySelector.children('option');
+      const input = element.querySelector('input');
+      if (!input) return;
+
+      input.value = formatTime(now);
+      const daySelector = document.getElementById(input.id.replace('time', 'day'));
+      const dayElements = daySelector.querySelectorAll('option');
```
Comment (weeman1337) [inline resources/assets/js/forms.js:130]: `daySelector` can possibly be `null`

File: resources/assets/js/forms.js [hunk 11]
```diff
@@ -163,7 +191,9 @@ $(() => {
...
   };
 
-  $('#collapseShiftsFilterSelect')
-    .on('hidden.bs.collapse', onChange)
-    .on('shown.bs.collapse', onChange);
+  document.getElementById('collapseShiftsFilterSelect')
+    ?.addEventListener('hidden.bs.collapse', onChange);
+
+  document.getElementById('collapseShiftsFilterSelect')
+    ?.addEventListener('shown.bs.collapse', onChange);
```
Comment (ximex) [inline resources/assets/js/forms.js:198]:
```js
const collapseShiftsFilterSelect = document.getElementById('collapseShiftsFilterSelect');
if (collapseShiftsFilterSelect) {
  const onChange = (e) => {
    localStorage.setItem('collapseShiftsFilterSelect', e.type);
  };

  collapseShiftsFilterSelect.addEventListener('hidden.bs.collapse', onChange);
  collapseShiftsFilterSelect.addEventListener('shown.bs.collapse', onChange);
}
```

File: resources/assets/js/forms.js [hunk 12]
```diff
@@ -128,15 +148,23 @@ $(function () {
...
+    });
   });
-  $('#oauth-settings-title').on('click', function () {
-    $('table tr.d-none').removeClass('d-none');
+
+  document.getElementById('oauth-settings-title')?.addEventListener('click', () => {
+    document.querySelectorAll('table tr.d-none').forEach((element) => {
+      element.classList.remove('d-none');
+    });
   });
```
Comment (ximex) [inline resources/assets/js/forms.js:168]:
```js
[
  ['welcome-title', '.btn-group.btn-group .btn.d-none'],
  ['settings-title', '.user-settings .nav-item'],
  ['oauth-settings-title', 'table tr.d-none'],
].forEach(([id, selector]) => {
  const titleElement = document.getElementById(id);
  if (titleElement) {
    titleElement.addEventListener('click', () => {
      document.querySelectorAll(selector).forEach((element) => {
        element.classList.remove('d-none');
      });
    });
  }
});
```

File: resources/assets/js/forms.js [hunk 13]
```diff
@@ -85,41 +97,49 @@ global.setHours = (hours) => {
   setInput(from, to);
 };
 
-$(function () {
+ready(() => {
   /**
      * Disable every submit button after clicking (to prevent double-clicking)
      */
```
Comment (ximex) [inline resources/assets/js/forms.js:103]: indentation (one space to much)

File: resources/assets/js/forms.js [hunk 14]
```diff
@@ -85,41 +97,49 @@ global.setHours = (hours) => {
...
-  $('form').submit(function (ev) {
-    $('input[type="submit"]').prop('readonly', true).addClass('disabled');
-    return true;
+  document.querySelectorAll('form').forEach((formElement) => {
+    formElement.addEventListener('submit', () => {
+      document.querySelectorAll('input[type="submit"],button[type="submit"]').forEach((element) => {
+        element.readOnly = true;
+        element.classList.add('disabled');
+      });
+      return true;
```
Comment (ximex) [inline resources/assets/js/forms.js:110]: necessary?

**Review by ximex — CHANGES_REQUESTED** [review]
this is all from my side

File: resources/assets/js/forms.js (hunk 12 above)
Comment (weeman1337) [reply resources/assets/js/forms.js:168]: Will be replaced with 

```
  [
    ['welcome-title', '.btn-group.btn-group .btn.d-none'],
    ['settings-title', '.user-settings .nav-item'],
    ['oauth-settings-title', 'table tr.d-none'],
  ].forEach(([id, selector]) => {
    document.getElementById(id)?.addEventListener('click', () => {
      document.querySelectorAll(selector).forEach((element) => {
        element.classList.remove('d-none');
      });
    });
  });

```

File: resources/assets/js/forms.js (hunk 14 above)
Comment (weeman1337) [reply resources/assets/js/forms.js:110]: :+1:

**Comment by weeman1337** [comment]
> you forgot some files: `sys_form.php`, `PublicDashboard_view.php`, `UserDriverLicenses_view.php`

Done

File: resources/assets/js/dashboard.js [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    if (!response.ok) {
+      console.warn('error loading dashboard');
+      return;
+    }
+
+    const responseData = await response.text();
+    const parser = new DOMParser();
+    const dummyDocument = parser.parseFromString(responseData, 'text/html');
+    const dashboardContent = dummyDocument.getElementById('public-dashboard');
+    document.querySelector('#content .wrapper').innerHTML = dashboardContent.innerHTML;
```
Comment (MyIgel) [inline resources/assets/js/dashboard.js:19]: The `#public-dashboard` div gets replaced which should not be the case

**Review by MyIgel — CHANGES_REQUESTED** [review]

**Pull request merged**